src/mongoose_rabbit_worker.erl

```diff
@@ -132,18 +135,34 @@ publish_groupchat_msg_recv(Channel, EventData = #{to_jid := To,
 publish_chat_msg_event(Channel, RoutingKey, #{from_jid := From,
                                               to_jid := To,
                                               msg := UserMsg,
-                                              exchange := Exchange}) ->
+                                              exchange := Exchange,
+                                              host := Host}) ->
     Message = make_chat_msg(From, To, UserMsg),
-    publish(Channel, Exchange, RoutingKey, Message).
+    publish(Channel, Exchange, RoutingKey, Message, Host).
 
 -spec publish(Channel :: pid(), Exchange :: binary(), RoutingKey :: binary(),
-              Message :: binary()) -> term().
-publish(Channel, Exchange, RoutingKey, Message) ->
-    amqp_channel:call(Channel,
-                      #'basic.publish'{
-                         exchange = Exchange,
-                         routing_key = RoutingKey},
-                      #amqp_msg{payload = Message}).
+              Message :: binary(), Host :: jid:server()) -> term().
+publish(Channel, Exchange, RoutingKey, Message, Host) ->
+    {PublishTime, Result} =
+        timer:tc(fun publish_message_and_wait_for_confirm/4,
+                 [Channel, Exchange, RoutingKey, Message]),
+    case Result of
+        true ->
+            mongoose_metrics:update(Host, ?MESSAGES_PUBLISHED_METRIC, 1),
+            mongoose_metrics:update(Host, ?MESSAGE_PUBLISH_TIME_METRIC,
+                                    PublishTime),
+            mongoose_metrics:update(Host, ?MESSAGE_PAYLOAD_SIZE_METRIC,
+                                    byte_size(Message)),
```

Comment:
I think this metrics updates may be hidden in separate function e.g:
```
update_metrics_message_published(Host, Message, PublishTime) ->
   mongoose_metrics:update(Host, ?MESSAGES_PUBLISHED_METRIC, 1),
   mongoose_metrics:update(Host, ?MESSAGE_PUBLISH_TIME_METRIC, PublishTime),
   mongoose_metrics:update(Host, ?MESSAGE_PAYLOAD_SIZE_METRIC, byte_size(Message)).
```

Reply:
or you can even earlier:
```
handle_message_publish(Result, Host, Message, PublishTime) ->
 ...
```
and call it instead of matching on `Result`

Reply:
You're right but I believe that it's only a matter of an approach a programmer chose :)